File: python/bulkpublish/templates.py
"""Post templates: saved text to start a new post from. Organization-wide, text only."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Dict, List, Optional

if TYPE_CHECKING:
    from .client import _BaseClient

from .types import PostTemplate


class TemplatesResource:
    """Operations on saved post templates (``client.templates``)."""

    def __init__(self, client: _BaseClient) -> None:
        self._client = client

    def list(self, *, kind: Optional[str] = None) -> List[PostTemplate]:
        """Templates in the organization, sorted by name.

        ``kind`` is ``caption`` (default) or ``first_comment``.
        """
        params = {"kind": kind} if kind is not None else None
        return self._client._request("GET", "/api/templates", params=params)["templates"]

    def get(self, template_id: int) -> PostTemplate:
        return self._client._request("GET", f"/api/templates/{template_id}")["template"]

    def create(self, *, name: str, content: str, kind: Optional[str] = None) -> PostTemplate:
        """Create a template. Up to 200 per organization per kind; names are unique per kind (default kind: ``caption``)."""
        body: Dict[str, Any] = {"name": name, "content": content}
        if kind is not None:
            body["kind"] = kind
        return self._client._request("POST", "/api/templates", json=body)["template"]

    def update(self, template_id: int, *, name: Optional[str] = None, content: Optional[str] = None) -> PostTemplate:
        body: Dict[str, Any] = {}
        if name is not None:
            body["name"] = name
        if content is not None:
            body["content"] = content
        return self._client._request("PUT", f"/api/templates/{template_id}", json=body)["template"]

    def delete(self, template_id: int) -> Dict[str, Any]:
        return self._client._request("DELETE", f"/api/templates/{template_id}")


class AsyncTemplatesResource:
    """Async variant of :class:`TemplatesResource`."""

    def __init__(self, client: _BaseClient) -> None:
        self._client = client

    async def list(self, *, kind: Optional[str] = None) -> List[PostTemplate]:
        params = {"kind": kind} if kind is not None else None
        return (await self._client._request("GET", "/api/templates", params=params))["templates"]

    async def get(self, template_id: int) -> PostTemplate:
        return (await self._client._request("GET", f"/api/templates/{template_id}"))["template"]

    async def create(self, *, name: str, content: str, kind: Optional[str] = None) -> PostTemplate:
        body: Dict[str, Any] = {"name": name, "content": content}
        if kind is not None:
            body["kind"] = kind
        return (await self._client._request("POST", "/api/templates", json=body))["template"]

    async def update(self, template_id: int, *, name: Optional[str] = None, content: Optional[str] = None) -> PostTemplate:
        body: Dict[str, Any] = {}
        if name is not None:
            body["name"] = name
        if content is not None:
            body["content"] = content
        return (await self._client._request("PUT", f"/api/templates/{template_id}", json=body))["template"]

    async def delete(self, template_id: int) -> Dict[str, Any]:
        return await self._client._request("DELETE", f"/api/templates/{template_id}")

File: node/src/templates.ts
import type { HttpClient } from './client.js';

export type TemplateKind = 'caption' | 'first_comment';

export interface PostTemplate {
  id: number;
  name: string;
  content: string;
  kind: TemplateKind;
  createdAt: string;
  updatedAt: string;
}

/**
 * Saved post templates — text to start a new post from. Text only: media,
 * channels and per-platform options belong to the post. `kind` picks between
 * a post caption and a saved first-comment reply; names are unique per kind.
 *
 * @example
 * ```ts
 * const { template } = await bp.templates.create({ name: 'Weekly roundup', content: 'This week:\n- ' });
 * await bp.posts.create({ content: template.content, channels: [...] });
 * ```
 */
export class TemplatesResource {
  constructor(private readonly http: HttpClient) {}

  /** List templates in the organization, sorted by name. Defaults to kind: 'caption'. */
  async list(params?: { kind?: TemplateKind }): Promise<{ templates: PostTemplate[] }> {
    const qs = params?.kind ? `?kind=${params.kind}` : '';
    return this.http.get(`/api/templates${qs}`);
  }

  async get(id: number): Promise<{ template: PostTemplate }> {
    return this.http.get(`/api/templates/${id}`);
  }

  /** Create a template. Up to 200 per organization; names are unique per kind (default kind: 'caption'). */
  async create(params: { name: string; content: string; kind?: TemplateKind }): Promise<{ template: PostTemplate }> {
    return this.http.post('/api/templates', params);
  }

  /** Rename and/or replace the text (at least one field). */
  async update(id: number, params: { name?: string; content?: string }): Promise<{ template: PostTemplate }> {
    return this.http.put(`/api/templates/${id}`, params);
  }

  async delete(id: number): Promise<{ success: boolean; deletedId: number }> {
    return this.http.delete(`/api/templates/${id}`);
  }
}
